src/vmm/src/devices/virtio/vsock/packet.rs

```diff
@@ -145,50 +236,90 @@ impl VsockPacket {
             return Err(VsockError::InvalidPktLen(hdr.len));
         }
 
-        if hdr.len > buffer.len() - VSOCK_PKT_HDR_SIZE {
+        if hdr.len > self.buffer.len() - VSOCK_PKT_HDR_SIZE {
             return Err(VsockError::DescChainTooShortForPacket(
-                buffer.len(),
+                self.buffer.len(),
                 hdr.len,
             ));
         }
+        self.hdr = hdr;
+        Ok(())
+    }
+
+    pub fn write_from_offset_to<T: WriteVolatile + Debug>(
+        &self,
+        dst: &mut T,
+        offset: u32,
+        count: u32,
+    ) -> Result<u32, VsockError> {
+        if count
+            > self
+                .buffer
+                .len()
+                .saturating_sub(VSOCK_PKT_HDR_SIZE)
+                .saturating_sub(offset)
+        {
+            return Err(VsockError::GuestMemoryBounds);
+        }
 
-        Ok(VsockPacket {
-            hdr,
-            buffer: VsockPacketBuffer::Tx(buffer),
-        })
+        self.buffer
+            .read_volatile_at(dst, (offset + VSOCK_PKT_HDR_SIZE) as usize, count as usize)
+            .map_err(|err| VsockError::GuestMemoryMmap(GuestMemoryError::from(err)))
+            .and_then(|read| read.try_into().map_err(|_| VsockError::DescChainOverflow))
     }
 
+    /// Returns the total length of this [`VsockPacket`]'s buffer (e.g. the amount of data bytes
+    /// contained in this packet).
+    ///
+    /// Return value will equal the total length of the underlying descriptor chain's buffers,
+    /// minus the length of the vsock header.
+    pub fn buf_size(&self) -> u32 {
+        self.buffer.len() - VSOCK_PKT_HDR_SIZE
+    }
+}
+
+/// Struct describing a single vsock packet.
+///
+/// Encapsulates the virtio descriptor chain containing the packet through the `IoVecBuffer[Mut]`
+/// abstractions.
+#[derive(Debug, Default)]
+pub struct VsockPacketRx {
+    /// A copy of the vsock packet's 44-byte header, held in hypervisor memory
+    /// to minimize the number of accesses to guest memory. Can be written back
+    /// to geust memory using [`VsockPacket::commit_hdr`] (only for RX buffers).
+    pub hdr: VsockPacketHeader,
+    /// The raw buffer, as it is contained in guest memory (containing both
+    /// header and payload)
+    buffer: IoVecBufferMut,
+}
+
+// This is needed because IoVecBufferMut contain *mut c_void type.
+// SAFETY:
+// Safe as only vmm thread will create and use this type.
+unsafe impl Send for VsockPacketRx {}
```

Comment:
maybe here we can instead just make `IoVecBufferMut` implement `Send`, to keep it consistent with `IoVecBuffer`

Reply:
done